### PR TITLE
Add devpi and devpi_consumer roles and devpi Docker container.

### PR DESCRIPTION
**Pull request by doctoryes**
Configuration Pull Request
---
https://openedx.atlassian.net/browse/PLAT-1912

devpi is a standalone caching layer in front of PyPI. This PR adds roles to:
- install devpi on an instance
- install pip configuration on an edxapp machine to "pip install" pkgs via the proxy instead of directly via PyPI

Also, adds a Dockerfile and supporting files to build a devpi Docker container that is used in Docker devstack.

Make sure that the following steps are done before merging:

  - [ ] A DevOps team member has approved the PR.
  - [ ] Are you adding any new default values that need to be overridden when this change goes live? If so:
    - [ ] Update the appropriate internal repo (be sure to update for all our environments)
    - [ ] If you are updating a secure value rather than an internal one, file a DEVOPS ticket with details.
    - [ ] Add an entry to the CHANGELOG.
  - [ ] If you are making a complicated change, have you performed the proper testing specified on the [Ops Ansible Testing Checklist](https://openedx.atlassian.net/wiki/display/EdxOps/Ops+Ansible+Testing+Checklist)?  Adding a new variable does not require the full list (although testing on a sandbox is a great idea to ensure it links with your downstream code changes).

### CODE REVIEW

File: docker/build/devpi/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+ARG ARG_DEVPI_SERVER_VERSION=4.4.0
+ARG ARG_DEVPI_WEB_VERSION=3.2.2
+ARG ARG_DEVPI_CLIENT_VERSION=4.0.0
+
+ADD . /edx/app/edx_ansible/edx_ansible
+
+WORKDIR /edx/app/edx_ansible/edx_ansible/docker/plays
+
+RUN echo '{ "allow_root": true }' > /root/.bowerrc
```
Comment (jmbowman) [inline docker/build/devpi/Dockerfile]: Does devpi actually use Bower?  I didn't see any mention of it in the docs or repo.

File: playbooks/roles/devpi/defaults/main.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+# Defaults for role devpi
+
+devpi_group: devpi
+devpi_user: devpi
+devpi_app_dir: /home/devpi
+devpi_venv_dir: devpi_venv
+devpi_venv_path: "{{ devpi_app_dir }}/venvs/{{ devpi_venv_dir }}"
+DEVPI_SERVER_VERSION: 4.4.0
+DEVPI_WEB_VERSION: 3.2.2
+DEVPI_CLIENT_VERSION: 4.0.0
```
Comment (jmbowman) [inline playbooks/roles/devpi/defaults/main.yml]: Do these really need to be specified both here and in the Dockerfile?  Seems like we'll forget to keep them in sync.

File: docker/build/devpi/docker-entrypoint.sh [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    export DEVPI_SERVERDIR DEVPI_CLIENTDIR
+}
+
+function initialize_devpi {
+    echo "[RUN]: Initializing devpi-server..."
+    devpi-server --restrict-modify root --start --host 127.0.0.1 --port 3141
+    devpi-server --status
+    devpi use http://localhost:3141
+    devpi login root --password=''
+    devpi user -m root password="${DEVPI_PASSWORD}"
```
Comment (jmbowman) [inline docker/build/devpi/docker-entrypoint.sh]: Is this defined in a separate secure repo?  On further reflection, I think we'd be fine randomizing and logging this; the other containers shouldn't need the password to download packages by default, and the user can fetch it from the log via a make target when they need to upload packages.

**Review by jmbowman — COMMENTED** [review]
Looks pretty good, just a few minor questions and suggestions.

File: playbooks/roles/devpi/defaults/main.yml (hunk 2 above)
Comment (doctoryes) [reply playbooks/roles/devpi/defaults/main.yml]: They don't - I just couldn't decide where to put the defaults. I'll remove them from here and leave them in the Dockerfile.

File: docker/build/devpi/Dockerfile (hunk 1 above)
Comment (doctoryes) [reply docker/build/devpi/Dockerfile]: Nope - this is copypasta. I will remove it.

File: docker/build/devpi/docker-entrypoint.sh (hunk 3 above)
Comment (doctoryes) [reply docker/build/devpi/docker-entrypoint.sh]: Currently, it's in the docker-compose.yml file in devstack, which isn't where it should go.
Where's the logic that randomizes the firefox/chrome VNC passwords upon container startup?

Comment (doctoryes) [reply docker/build/devpi/docker-entrypoint.sh]: Nm - I implemented a solution that generates a random password upon the first initialization of the devpi volumes. The password gets stashed into a file in the volume and that file persists across container start/stops, which mirrors the persistence of the root password. Let me know what you think about it.

**Review by jmbowman — APPROVED** [review]

**Comment by doctoryes** [comment]
@edx/devops I built a sandbox with this configuration branch - it built successfully fyi.

File: docker/build/devpi/ansible_overrides.yml [hunk 4]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (feanil) [inline docker/build/devpi/ansible_overrides.yml]: Is the empty dict necessary?

File: playbooks/edx-east/edxapp.yml [hunk 5]
```diff
@@ -11,6 +11,8 @@
      when: COMMON_ENABLE_AWS_ROLE
    - role: automated
      AUTOMATED_USERS: "{{ EDXAPP_AUTOMATED_USERS | default({}) }}"
+   - role: devpi_consumer
```
Comment (feanil) [inline playbooks/edx-east/edxapp.yml]: Since devstacks are now docker do you need this if you have it in the edxapp.yml in the docker directory?

**Review by feanil — COMMENTED** [review]
I didn't review the role, just the integration points and had some questions.

File: docker/build/devpi/ansible_overrides.yml (hunk 4 above)
Comment (doctoryes) [reply docker/build/devpi/ansible_overrides.yml]: This override file isn't used at all - but an empty dict was required to include it in the ansible-playbook call. I'll just remove it.

File: playbooks/edx-east/edxapp.yml (hunk 5 above)
Comment (doctoryes) [reply playbooks/edx-east/edxapp.yml]: Right - seems like the `when` will never be true. I'll remove it from this location.

**Comment by doctoryes** [comment]
@feanil I responded to your feedback. Re-review?

**Comment by doctoryes** [comment]
@edx/devops Though this work isn't mission-critical, this PR is blocked on a DevOps review. And devpi can't be added to devstack until this PR merges - FYI.

**Comment by coryleeio** [comment]
This is a pretty big PR and I can't look at it just yet, but i've gone
ahead and retagged all of devops on it so it will come up in our review
queue so it doesn't get lost

**Review by coryleeio — APPROVED** [review]

**Pull request merged**